kvbc/src/v4blockchain/v4_blockchain.cpp

```diff
@@ -0,0 +1,116 @@
+// Concord
+//
+// Copyright (c) 2022 VMware, Inc. All Rights Reserved.
+//
+// This product is licensed to you under the Apache 2.0 license (the
+// "License").  You may not use this product except in compliance with the
+// Apache 2.0 License.
+//
+// This product may include a number of subcomponents with separate copyright
+// notices and license terms. Your use of these subcomponents is subject to the
+// terms and conditions of the subcomponent's license, as noted in the LICENSE
+// file.
+
+#include "v4blockchain/v4_blockchain.h"
+#include "v4blockchain/detail/utils.h"
+#include "categorization/base_types.h"
+#include "categorization/db_categories.h"
+#include "kvbc_key_types.hpp"
+#include "block_metadata.hpp"
+
+namespace concord::kvbc::v4blockchain {
+
+using namespace concord::kvbc;
+
+KeyValueBlockchain::KeyValueBlockchain(
+    const std::shared_ptr<concord::storage::rocksdb::NativeClient>& native_client,
+    bool link_st_chain,
+    const std::optional<std::map<std::string, categorization::CATEGORY_TYPE>>& category_types)
+    : native_client_{native_client},
+      block_chain_{native_client_},
+      state_transfer_chain_{native_client_},
+      latest_keys_{native_client_, category_types} {}
+
+/*
+  1 - check if we can perform GC on latest CF
+  2 - add the block to the blocks CF
+  3 - add the keys to the latest CF
+  4 - atomic write to storage
+  5 - increment last reachable block.
+*/
+BlockId KeyValueBlockchain::add(categorization::Updates&& updates) {
+  auto scoped = v4blockchain::detail::ScopedDuration{"Add block"};
+  // Should be performed before we add the block with the current Updates.
+  auto sequence_number = markHistoryForGarbageCollectionIfNeeded(updates);
```

Comment:
nit: Can we do markHistoryForGarbageCollectionIfNeeded in a threadpool.

Reply:
it can be, let's measure it and see if it's needed.